src/sentry/reprocessing2.py

```diff
@@ -0,0 +1,178 @@
+    return hashlib.md5(
+        u"{}:{}:unprocessed".format(project_id, event_id).encode("utf-8")
+    ).hexdigest()
+
+
+def save_unprocessed_event(project, event_id):
+    """
+    Move event from event_processing_store into nodestore. Only call if event
+    has outcome=accepted.
+    """
+    if not features.has("projects:reprocessing-v2", project, actor=None):
+        return
+
+    with sentry_sdk.start_span(
+        op="sentry.reprocessing2.save_unprocessed_event.get_unprocessed_event"
+    ):
+        data = event_processing_store.get(
+            cache_key_for_event({"project": project.id, "event_id": event_id}), unprocessed=True
+        )
+        if data is None:
+            return
+
+    with sentry_sdk.start_span(op="sentry.reprocessing2.save_unprocessed_event.set_nodestore"):
+        node_id = _generate_unprocessed_event_node_id(project_id=project.id, event_id=event_id)
+        nodestore.set(node_id, data)
+
+
+def backup_unprocessed_event(project, data):
+    """
+    Backup unprocessed event payload into redis. Only call if event should be
+    able to be reprocessed.
+    """
+
+    if not features.has("projects:reprocessing-v2", project, actor=None):
+        return
+
+    event_processing_store.store(data, unprocessed=True)
+
+
+def delete_unprocessed_events(events):
+    node_ids = [
+        _generate_unprocessed_event_node_id(event.project_id, event.event_id) for event in events
+    ]
+    nodestore.delete_multi(node_ids)
+
+
+def reprocess_event(project_id, event_id, start_time):
+    node_id = _generate_unprocessed_event_node_id(project_id=project_id, event_id=event_id)
+
+    with sentry_sdk.start_span(op="reprocess_events.nodestore.get"):
+        data = nodestore.get(node_id)
+
+    from sentry.tasks.store import preprocess_event_from_reprocessing
+    from sentry.ingest.ingest_consumer import CACHE_TIMEOUT
+
+    # Take unprocessed data from old event and save it as unprocessed data
+    # under a new event ID. The second step happens in pre-process. We could
+    # save the "original event ID" instead and get away with writing less to
+    # nodestore, but doing it this way makes the logic slightly simpler.
+
+    # Step 1: Fix up the event payload for reprocessing and put it in event
+    # cache/event_processing_store
+    tags = dict(data.get("tags") or ())
+    orig_event_id = tags["original_event_id"] = data["event_id"]
+    event = eventstore.get_event_by_id(project_id, orig_event_id)
+    if event is None:
+        return
+
+    tags["original_group_id"] = event.group_id
+    data["tags"] = list(tags.items())
+
+    event_id = data["event_id"] = uuid.uuid4().hex
+
+    # XXX: Only reset received
+    data["timestamp"] = data["received"] = start_time
+
+    cache_key = event_processing_store.store(data)
+
+    # Step 2: Copy attachments into attachment cache
+    queryset = EventAttachment.objects.filter(
+        project_id=project_id, event_id=orig_event_id
+    ).select_related("file")
```

Comment:
In a follow-up: We can consider to only load attachments required for processing, that is, crash reports. For other attachments, we can do one of:

1. Copy file records in save_event without accessing blob store, if we keep duplicating the events.
2. Do nothing, if we choose to mutate events.

Reply:
if we just copy the file records won't blobstore have to have a reference count such that the blob is deleted on the second event deletion, not the first?

Reply:
This is how blob store works. There is an association between files and blobs, so that you can clone files without re-uploading their blobs. Have a look at [`FileBlobIndex`](https://github.com/getsentry/sentry/blob/bcab979e02fc78c6b24895f94cb7a46762c622cd/src/sentry/models/file.py#L435-L436) if you're interested.